### ISSUE TITLE
Memory space can only be set statically

### ISSUE THREAD

**Issue by ashermancinelli** [LLNL/hiop#131]
Currently, `LinearAlgebraFactory` class is managing HiOp memory space using static methods. Using static methods may be inflexible solution that could potentially limit future development, especially on hybrid hardware architectures.

**Comment by pelesh** [comment]
Most likely using singleton pattern would provide enough flexibility for implementations on heterogeneous hardware. I would consider that first.

**Comment by ashermancinelli** [comment]
I don't think Singleton pattern would give us anything over our current api. I believe we discussed using factory pattern offline though, is this correct?

**Comment by cnpetra** [comment]
the linear algebra factory logic should be part of NlpFormulation (and it is right now to some extent) as this is the entity that needs to be assigned and ran into a particular memory space. Underneath the NlpFormulation,  the code decides memory space and types for linear algebra objects based on the user options; then uses `new_clone` and `alloc_copy` to create new objects. However, in some parts, `new_clone` and `alloc_copy` are not sufficient, as vectors/matrices of new sizes need to be created. Also, sometimes the vectors/matrices need to be in a different memory space (usually host) than the NlpFormulation is supposedly running in. I think singleton pattern would work for these situations. Factory pattern can be a bit more invasive because the specific instance of the linear algebra factory needs to be available to all the inner classes creating vectors/matrices; also, it needs to be able allocate in different memory spaces, which kinda defeats the purpose. 

In other words, NlpFormulation is already a factory pattern and makes more sense for it to rely on a  a singleton linear algebra factory.

**Comment by pelesh** [comment]
Provided #241 works out well, we could get what we need with minimal changes to the existing linear algebra factory and have something like:
```c++
auto& f = hiopLinAlgFactory::getInstance(); 
hiopVector* x = f.createVector<hiopCudaPolices>(1000, "UM");
```
This would use singleton implementation to create a RAJA vector with data in unified memory. Probably there are some details to sort out, but this could work for us.

**Comment by ashermancinelli** [comment]
What does this give us over static method?
```cpp
auto* vec = hiopLinAlg::creatVector<hiopCudaPolicies>(size, mem_space);
```

**Comment by cnpetra** [comment]
> What does this give us over static method?
> 
> ```c++
> auto* vec = hiopLinAlg::creatVector<hiopCudaPolicies>(size, mem_space);
> ```

It appears that static-only methods do the job for now (meaning fixed policies), see PR#274

**Comment by cnpetra** [comment]
closed by PR#274
